src/primitives/transaction.h

```diff
@@ -351,11 +357,39 @@ class CTransaction
 
     template <typename Stream, typename Operation>
     inline void SerializationOp(Stream& s, Operation ser_action, int nType, int nVersion) {
-        READWRITE(*const_cast<int32_t*>(&this->nVersion));
+        if (ser_action.ForRead()) {
+            // When deserializing, unpack the 4 byte header to extract fOverwintered and nVersion.
+            uint32_t header;
+            READWRITE(header);
+            *const_cast<bool*>(&fOverwintered) = header >> 31;
+            *const_cast<int32_t*>(&this->nVersion) = header & 0x7FFFFFFF;
+        } else {
+            if (fOverwintered) {
+                // When serializing Overwintered tx, assemble the 4 byte header from fOverwintered and nVersion
+                uint32_t header = (1 << 31) | this->nVersion;
+                READWRITE(header);
+            } else {
+                // When serializing v1 and v2, the 4 byte header is nVersion
+                READWRITE(*const_cast<int32_t*>(&this->nVersion));
+            }
+        }
         nVersion = this->nVersion;
+        if (fOverwintered) {
+            READWRITE(*const_cast<uint32_t*>(&this->nFormatBranchId));
+        }
+        // FIXME: Do we want to do this?  Will it lead to unexpected code paths?
+        if (fOverwintered && nVersion != 3) {
```

Comment:
Check `nFormatBranchId` here (probably using the same `tx.nVersion -> tx.nFormatBranchId` mapping as in the consensus rules).

Reply:
Updated to check version group id too.